app/src/main/java/org/mozilla/fenix/components/toolbar/BrowserMenuSignIn.kt

```diff
@@ -0,0 +1,77 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.toolbar
+
+import android.content.Context
+import android.view.View
+import android.widget.TextView
+import androidx.annotation.ColorRes
+import androidx.annotation.DrawableRes
+import androidx.annotation.VisibleForTesting
+import androidx.core.content.ContextCompat
+import mozilla.components.browser.menu.BrowserMenu
+import mozilla.components.browser.menu.item.BrowserMenuImageText
+import org.mozilla.fenix.R
+import org.mozilla.fenix.ext.components
+
+/**
+ * A menu item for displaying account information. The item calculates the label on every bind call,
+ * to provide each menu with the latest account manager state.
+ *
+ * @param label The function that provides the visible label of this menu item.
+ * @param imageResource ID of a drawable resource to be shown as icon.
+ * @param textColorResource Optional ID of color resource to tint the text.
+ * @param listener Callback to be invoked when this menu item is clicked.
+ */
+
+class BrowserMenuSignIn(
+    @ColorRes private val textColorResource: Int,
+    @DrawableRes imageResource: Int = R.drawable.ic_signed_out,
+    label: String = String(),
+    listener: () -> Unit = {},
+) : BrowserMenuImageText(
+    label,
+    imageResource,
+    textColorResource = textColorResource,
+    listener = listener,
+) {
+    override fun bind(menu: BrowserMenu, view: View) {
+        super.bind(menu, view)
+        val textView = view.findViewById<TextView>(R.id.text)
+        textView.text = getLabel(textView.context)
+        textView.setTextColor(ContextCompat.getColor(textView.context, textColorResource))
+    }
+
+    /**
+     * Determines the proper label for the sign in button.  There are three states:
+     * user is signed-in, please sign-in, and we are not sure what's happening. The last state does
+     * occur when the menu is opened before the AccountManager has finished updating its state, or
+     * when there is no connection and AccountManager doesn't return us the profile because its cache
+     * has 2 minutes 'freshness':
+     * https://searchfox.org/mozilla-mobile/source/application-services/components/fxa-client/src/internal/profile.rs#41
+     */
+    @VisibleForTesting
+    internal fun getLabel(context: Context): String {
+        return with(context) {
+            val isAuthenticated = components.settings.signedInFxaAccount
+            val isAccountManagerReady = components.backgroundServices.accountManagerAvailableQueue.isReady()
+            val email = components.backgroundServices.syncStore.state.account?.email
+
+            if (isAccountManagerReady) {
+                if (isAuthenticated) {
```

Comment:
@Mugurell I remember discussing with you about adding `enum` for the FXA sign in states that are available so we wouldn't have to do all this if else checks and can easily encapsulate the states in a class. Do you happen to remember where that is and if we could reuse that here?

Reply:
Good memory!
Yeah, we have a [FenixAccountManager](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/java/org/mozilla/fenix/components/accounts/FenixAccountManager.kt) which already contains the `accountState` and `accountProfileEmail` info.
While the `FxaAccountManager` that is wrapped in `FenixAccountManager` has a lengthy initialization it seems like it is called in `FenixApplication` also so it should be ready by the time the menu is opened.

Reply:
@mavduevskiy Could you look at using `FenixAccountManager`? This should also help simplify the lengthy comment you have here because the account state will already be documented (unless the freshness state is not accounted there then we should just make sure that state is also documented).

Reply:
@gabrielluong 
One of the main goals is to be able to accommodate the `The account manager needs some time to get updated` case, but `FenixAccountManager` reports `no_account` state in that case. `no_account` and `no_information_yet` are different UI cases: we ask to sign in in the first case, and show general "account info" in the second. 
As for the `profile` information and `is_signed_in` state I went with @MatthewTighe [suggestion](https://github.com/mozilla-mobile/fenix/pull/28409#discussion_r1067545553).

About the comment, I didn't find any documentation or logic to accommodate the state when we have an account but not a profile, hence the comment on why in the `isAuthenticated` we still want to check if the profile is not null. Should be able to remove it once the [caching mechanism](https://github.com/mozilla/application-services/pull/5333) gets fully available.

Reply:
> While the `FxaAccountManager` that is wrapped in `FenixAccountManager` has a lengthy initialization it seems like it is called in `FenixApplication` also so it should be ready by the time the menu is opened.

We introduced the SyncStore that holds the account state (see: [SyncStatus](https://searchfox.org/mozilla-mobile/source/firefox-android/android-components/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/store/SyncState.kt)) as well to have one place where you can know this information that also follows our store pattern.

The hope is that we can remove the FenixAccountManager and other duplications as well so we can just query with `syncStore.state.status`.

Reply:
It sounds like we have another approach to investigate with the `SyncStore`. I don't have the necessary bandwidth to look into this too deeply right now, but it would be good to know what you decide on with all this new information.

Reply:
This is well beyond the scope of this PR at this point, but I wanted to address my thoughts in context:

If the enums defined in `FenixAccountManager` are a desirable method of determining the state of account state then I think it would be preferable to add them to the `SyncStore`, in order to continue simplifying Sync code. Looking `FenixAccountManager`, I don't see any reason why we wouldn't want something similar in the `SyncStore`. We could even investigate an update to `SyncState` such that we had something like:

```
data class SyncState(
    val accountState: AccountState? = null,
) : State

sealed class AccountState {
    data class Authenticated(val account: Account) : AccountState()
    object NoAccount : AccountState()
    object NeedsReauth : AccountState()
}
```

which seems better anyway to me than needing intrinsic knowledge about `SyncState.account`.

> What I am implicitly looking for is some form of simplification similar to FenixAccountManager to all of this if else checks to get the account state and have it broadly documented and be reusable since this has been a common problem and these checks are still spread out in various parts of the code.

The above I think would address this as well as we are able, but does not deal with one of the underlying issues - the only mechanism we have for caching whether users are authenticated is through `Settings`. I also think it would be worth adding that cache to A-C, but I am not sure what that would look like

Reply:
In general, the `SyncStore` is meant to address the fundamental problem that @mavduevskiy has described in the body of the PR. There are a couple options listed there, but another is to use reactive programming. That is the intent of the observable Stores, but I agree with @mavduevskiy that overriding the `bind` function here is a good choice because it is a smaller lift than making starting to make this Toolbar use reactive programming. If this were to be updated to Compose I think that would be a good time to make a more fundamental change like that.